Penalties Rewrite (#309)
helium/docs: docs/mine-hnt/validators/validator-score.mdx

A penalty of `1.0` is added at the end of each consensus group for each Validator that participated in that group.  If the Validator is not ejected from the consensus group, it will receive another `1.0` penalty at the end of the 2nd consensus group, now having a score of `2.0`.  

This penalty of `1.0` is added even if the Validator is ejected from the consensus group as it is a penalty added for participating in the ending consensus group. This penalty is called the `tenure_penalty`.

### DKG (Distributed Key Generation) Penalty
* This penalty can occur if an election fails, and the Validator was part of the reason for the election to fail.
* At the end of the election, all Validators vote and sign the election result.  If the election fails, Validators that a majority agree did not sign the election results are penalized.
* This penalty is applied once per consensus.
* The amount applied is a chain variable, currently set at `1.0`. The variable is viewable by running `miner ledger variables dkg_penalty`.

> * Learn more about [Distributed Key Generation](https://en.wikipedia.org/wiki/Distributed_key_generation)

Reviewer: This bullet point would likely flow better as a "Info" box.

* This penalty can occur if a majority of other Validators in the consensus group report your Validator as not being "seen"
* This is equivalent to the Validator being offline, not available, or not communicating with the group.
* This penalty is applied for each occurrence, which can happen multiple times during a consensus group.
* The amount applied per occurrence is a chain variable, currently set at `0.33`. The variable is viewable by running `miner ledger variables election_seen_penalty`.

**BBA (Byzantine Binary Agreement) Penalties**
* This penalty can occur when the Validator is seen but fails to contribute a bundle of transactions and metadata to the block.  
* Acquiring this penalty usually means the Validator is slow/underpowered, or poorly connected.
* This may also occur if the time on your Validator is skewed vs the group time.  [Run `ntp`](https://en.wikipedia.org/wiki/Network_Time_Protocol)!
* This penalty is applied for each occurrence, which can happen multiple times during a consensus group.
* The amount applied per occurrence is a chain variable, currently set at `0.1`. The variable is viewable by running `miner ledger variables election_bba_penalty`.

> * Learn more about [Byzantine Agreements](https://en.wikipedia.org/wiki/Byzantine_fault) and their affect on [Consensus](https://en.wikipedia.org/wiki/Consensus_(computer_science))

Reviewer: Same as above. This would be better a "Info" box.

## Penalty Decay

Penalties decay over time.  

Each penalty, when assessed, also decays at the same rate.  This decay rate is a chain variable, able to be viewed by running `miner ledger variables penalty_history_limit`.  

> Every time a penalty is assessed, it begins to decay at `penalty * (1 - age/penalty_history_limit)`, where age is the # of blocks that have occurred since the penalty was put in place.  This means that as blocks are formed, each penalty decays over time, bringing a Validator back down to `0.0`.

Reviewer: ```suggestion
Every time a penalty is assessed, it begins to decay at `penalty * (1 - age/penalty_history_limit)`, where age is the number of blocks that have occurred since the penalty was put in place.  This means that as blocks are formed, each penalty decays over time, bringing a Validator back down to `0.0`.
```

## Penalty Objectives

Through these rules, the penalties help maintain the the goals of the Validator network:
* Ensuring that poor performing Validators have a less favorable chance of being included in a consensus group
* Ensuring that poor performing Validators have an increased chance of being ejected in an election
* A more even distribution of rewards by increasing a Validator's chance for ejection as they participate in sequential consensus groups

## Viewing Your Penalties

> There are a couple tools available to you to view the current penalties of your validator, all validators, or ones currently elected to the consensus group.

Reviewer: ```suggestion
There are a several tools available to you to view the current penalties of your validator, all validators, or ones currently elected to the consensus group.
```

**Viewing the Ledger**

> Metadata about validators, as well as current penalty data is available in the ledger. View the ledger by running the following command:

Reviewer: ```suggestion
Metadata about validators, as well as current penalty data, is available in the ledger. View the ledger by running the following command:
```

---
id: validator-score
title: Validator Penalties and Impact on Rewards
hide_title: true
sidebar_label: Penalties
slug: /mine-hnt/validators/validator-score
---

import useBaseUrl from "@docusaurus/useBaseUrl";

# Validator Penalties and Impact on Rewards

Validators are a new entity on the Helium Blockchain that will perform the work of the consensus group including verifying transactions and adding new blocks to the blockchain.  As Validators will play an important role in the expansion, stability and success of the Helium network moving forward, it's critical that community run Validators meet the performance requirements of their role.

The consensus group is made up of Validator nodes that are rewarded with $HNT for working together to verify transactions and producing blocks. 

> However, if a Validator node prevents the consensus group from performing its task, for example, uses insufficient computing resources or a weak network connection (equally strong upstream and downstream is needed), the Validator node will incur penalties that can affect the amount of rewards earned.

Reviewer 2: Minor grammatical error here. Possibly want something like this:

> However if a Validator node prevents the consensus group from performing its task, for example if it provides insufficient computing resources or has a weak network connection (equally strong upstream and downstream is needed), the Validator node will incur penalties that can affect the amount of rewards earned.

**The ejected 25% is replaced by available Validators:**  
* This is also probabilistic, based on the current penalty scores of the Validators available to be included in the consensus group.
* Validators with _higher_ penalty score will have a _lower_ chance of being elected
* Validators with _lower_ penalty score will have a _higher_ chance of being elected

It's important to understand that elections are still based on probability. Merely having a high penalty will not ensure a Validator is removed from a consensus group. By contrast, having a low penalty score does not ensure that a Validator will _remain_ in a consensus group. 

Upon joining the Network, Validator nodes start with a base penalty of `0.0`. 

## Penalty Types

Now that there's a high-level description of how penalties affect the election process, let's define how penalties are assessed:

### Consensus Group Participation

To ensure that consensus groups members are continually churned, and that no individual Validator participates for too long, penalties are added. This is not a performance based penalty and is added to _all_ Validators that participate in a consensus group.

> A penalty of `1.0` is added at the end of each consensus group for each Validator that participated in that group.  If the Validator is not ejected from the concensus group, it will receive another `1.0` penalty at the end of the 2nd concensus group, now having a score of `2.0`.  

Reviewer 3: "consensus" (not "concensus")

This penalty of `1.0` is added even if the Validator is ejected from the consensus group as it is a penalty added for participating in the ending consensus group.

### Performance Based

The details of the HoneyBadgerBFT protocol used is out of scope for this document, but this can be read more about [here](https://eprint.iacr.org/2016/199.pdf).  What's important is that there are a few different types of failures that will create performance penalties that can be applied to a Validator while in consensus:

**Seen Penalties**
* This penalty can occur if a majority of other Validators in the consensus group report your Validator as not being "seen"

> * This is equivalent to the Validator being offline, not available, or not responding to the group.

Reviewer 4: ```suggestion
* This is equivalent to the Validator being offline, not available, or not communicating with the group.
```

> * This is a penalty of `0.33` for each occurance

Reviewer 3: occurrence

> * This is a penalty of `0.33` for each occurrence

Reviewer 5: I think this is worded a bit awkwardly. How about something like this (would apply to this and BBA section)

* This penalty is applied for each occurrence, which can happen multiple times during an consensus group.
* The amount applied per occurrence is a chain variable, currently set at 0.33. The variable is viewable by running `miner ledger variables election_seen_penalty`

> * This penalty is applied for each occurrence, which can happen multiple times during an consensus group.

Reviewer 6: [nit]
```suggestion
* This penalty is applied for each occurrence, which can happen multiple times during a consensus group.
```

* The amount applied per occurrence is a chain variable, currently set at `0.33`. The variable is viewable by running `miner ledger variables election_seen_penalty`.

> **BBA Penalties**

Reviewer 6: [question] what does BBA stand for?

Reviewer 4: Byzantine Binary Agreement

Author: Added to Doc with rewrite

* This penalty can occur when the Validator is seen but fails to contribute a bundle of transactions and metadata to the block.  
* Acquiring this penalty usually means the Validator is slow/underpowered, or poorly connected.

> * this is a penalty of `0.1` for each occurance

Reviewer 3: occurrence

> * Acquiring this penalty usually means the Validator is slow/underpowered, or poorly connected.

Reviewer 4: Significant clock skew may also be a factor here.

> * This penalty is applied for each occurrence, which can happen multiple times during an consensus group.

Reviewer 6: [nit]
```suggestion
* This penalty is applied for each occurrence, which can happen multiple times during a consensus group.
```

* The amount applied per occurrence is a chain variable, currently set at `0.1`. The variable is viewable by running `miner ledger variables election_bba_penalty`.

> **DKG Penalties**

Reviewer 6: [question] what does DKG stand for?

Reviewer 4: Distributed Key Generator

* This penalty can occur if an election fails, and the Validator was part of the reason for the election to fail.
* At the end of the election, all Validators vote and sign the election result.  If the election fails, Validators that a majority agree did not sign the election results are penalized.
* This is a penalty of `1.0` for each occurrence

> * Thiis is a chain variable, able to be viewed by running `miner ledger variables dkg_penalty`

Reviewer 5: nit: s/Thiis/This/
